src/ORM/Hierarchy/Hierarchy.php

```diff
@@ -333,39 +333,13 @@ public static function prepopulate_numchildren_cache($baseClass, $stage, $idList
         }
 
         /** @var Versioned|DataObject $singleton */
-        $dummyObject = new $baseClass();
-        $dummyObject->ID = -23478234; // going to use this for search & replace
+        $dummyObject = DataObject::singleton($baseClass);
         $baseTable = $dummyObject->baseTable();
 
         $idColumn = Convert::symbol2sql("{$baseTable}.ID");
-        $parentIdColumn = Convert::symbol2sql("ParentID");
-
 
         // Get the stageChildren() result of a dummy object and break down into a generic query
-        $query = $dummyObject->stageChildren(true)->dataQuery()->query();
-        $where = $query->getWhere();
-
-        $newWhere = [];
-
-        foreach ($where as $i => $compoundClause) {
-            foreach ($compoundClause as $clause => $params) {
-                // Skip any "WHERE ParentID = [marker]" clauses as this will be replaced with a GROUP BY
-                if (!(preg_match('/^"[^"]+"\."ParentID" = \?$/', $clause) && $clause[1] == $dummyObject->ID)) {
-                    // Replace any marker IDs with "<basetable>"."ID"
-                    $paramNum = 0;
-                    foreach ($params as $j => $param) {
-                        if ($param == $dummyObject->ID) {
-                            unset($params[$j]);
-                            $clause = DB::replace_parameter($clause, $paramNum, $parentIdColumn, true);
-                        } else {
-                            $paramNum++;
-                        }
-                    }
-
-                    $newWhere[] = [ $clause => $params ];
-                }
-            }
-        }
+        $query = $dummyObject->stageChildren(true, true)->dataQuery()->query();
```

Comment:
We can probably stop the implementation of DB::replace_parameter if we go with this, since it's a bit gross and was only added for this use-case.

Reply:
We need to add "Where ID != ParentID" as a filter here.

Reply:
I've added the `Where ID != ParentID` statement to stageChildren.